Utils/Dataflow/030_PDFAnalyzer/xmltable.py

```diff
@@ -367,46 +370,110 @@ def analyze_page(text):
     rows.sort(key=lambda row: row_centery(row))
+    # Discard the top row if it contains "DRAFT" line. Such line is sometimes
+    # placed on top of each page of the document and may interfere with
+    # reconstruction when the caption is below the tables.
+    for line in rows[0]:
+        if line.text == "DRAFT":
+            rows = rows[1:]
+            break
 
     return rows
 
 
 def get_tables_from_text(text):
     """ Get tables from a xml page text. """
-    re_textbox = re.compile(r"<textbox id=\"\d+\" bbox=\"([0-9.,]+)\">",
-                            re.DOTALL)
-    re_table_caption = re.compile(r"Table \d+:")
-    tlines = re_textline.findall(text)
-    lines = []
-    table_captions = []
-    for line in tlines:
-        tl = TextLine(line)
-        if re_table_caption.match(tl.text):
-            table_captions.append(tl)
+    rows = analyze_page(text)
+    caption_start = False
+    caption_rows = []
+    text_rows = []
+    # Determine the type of each row, if possible.
+    # Caption rows are related to the table - notably, they contain the table
+    # number.
+    # Text rows contain parts of the document's main text body.
+    # The tables cannot include rows from any of these two types - therefore,
+    # such rows (and document borders) can be used to determine whether the
+    # table captions are positioned above or below their tables.
+    for (i, row) in list(enumerate(rows)):
+        if row[0].text.startswith("Table ") and\
+           re_table_caption_short.match(row[0].text):
+            caption_start = row
+            caption_rows.append(i)
+        elif caption_start and len(row) == 1 and\
+                abs(row[0].left - caption_start[0].left) < 1.0:
+            caption_rows.append(i)
         else:
-            lines.append(tl)
-
-    # Find the highest top coordinate possible and use it as a zero
-    # point for new Y axis.
-    top = max(table_captions + lines, key=lambda line: line.top).top
-    for line in table_captions + lines:
-        line.swap_y(top)
-
-    table_captions.sort(key=lambda x: x.center[1])
-
-    table_lines = []
+            caption_start = False
+            if len(row) > 1 and row[0].right - row[0].left < 10.0:
+                f = row[1]
+            else:
+                f = row[0]
+            # Magic number - text rows in most of the documents are positioned
+            # in such way that PDFMiner determines their left coordinate to
+            # be ~76. Non-rotated pages only.
+            if f.left - 76.0 < 1.0 and len(row) < 3:
+                text_rows.append(i)
+    # Tables' captions position on the page
+    # -1 - above tables
+    # 0 - unknown
+    # 1 - below tables
+    caption_position = 0
+    if 0 in caption_rows:
+        caption_position = -1
+    elif len(rows) - 1 in caption_rows:
+        caption_position = 1
+    else:
+        for i in caption_rows:
+            if i - 1 in text_rows:
+                caption_position = -1
+                break
```

Comment:
Because there can be more than one table on the page. 
To be precise, in each group of caption rows the rows next to first and last must be checked. For example, if we have the following caption rows:
[2, 3, 4, 8, 9],
then we must check:
[1, 5, 7, 10] = [2-1, 4+1, 8-1, 9+1].
Why bother with separating "border caption rows" from the other ones and determining which one is "first" and which is "last" when we can check i+-1 for each one? I doubt that the former will be faster/easier to code.

Reply:
Right, I missed this moment.

Hmm, then maybe it would be more readable (here and in the following code, where it is also required to skip rows belonging to same caption) if make `caption_rows` two-dimentional? Like `caption_rows[<caption_id>] = [<caption_row1>, <caption_row2>, ...]`.

I do not say it should be done right now, just an idea for the future. It is a bit painful to follow the `i+-1` logic, when in the loop we need to run the page rows back and forth... for me, at least ;) But if you are OK with it and never feel like loosing the context (even getting back to this code a year from now) -- that\`s really good for you.

You may be right saying:
> I doubt that the former will be faster/easier to code.

But, as they say: code is read more often than it is written. So it worth spending time on readability improvement, to gain from it much more time in the future for yourself (and no only yourself, in fact).

Reply:
> Hmm, then maybe it would be more readable (here and in the following code, where it is also required to skip rows belonging to same caption) if make `caption_rows` two-dimentional? Like `caption_rows[<caption_id>] = [<caption_row1>, <caption_row2>, ...]`.

Seems reasonable, I'll look into it during future development.